documentation/projects/proposals/dark_mode/20240325-implementation_plan_dark_mode.md

```diff
@@ -0,0 +1,263 @@
+<!-- This template is exhaustive and may include sections which aren't relevant to your project. Feel free to remove any sections which would not be useful to have. -->
+
+## Reviewers
+
+<!-- Choose two people at your discretion who make sense to review this based on their existing expertise. Check in to make sure folks aren't currently reviewing more than one other proposal or RFC. -->
+
+- [ ] @obulat
+- [ ] @sarayourfriend
+
+## Project links
+
+<!-- Enumerate any references to other documents/pages, including milestones and other plans -->
+
+- [Project Thread](https://github.com/WordPress/openverse/issues/3592)
+- [Project Proposal](./20240313-project_proposal_dark_mode.md)
+
+## Overview
+
+<!-- An overview of the implementation plan, if necessary. Save any specific steps for the section(s) below. -->
+
+This dark mode implementation plan is comprised of three work streams:
+
+- Color management in Tailwind and Vue frontend components
+- Toggling dark mode (new UI and dark mode detection logic)
+- Visual regression tests and feature flagging
+
+Most of these work streams can happen in parallel which I will elaborate on in
+the ["Implementation"](#implementation) section.
+
+### Design Philosophy
+
+Understanding the way @fcoveram has designed the color system for dark mode is
+crucial to understanding this implementation. Quite simply and elegantly, the
+designs use a "palette swap" approach in which each color has a 1:1 replacement
+from light mode to dark mode.
+
+![Openverse.org dark mode color palette](/_static/dark_mode_palette_example.png)
+
+While we _will_ include easy mechanisms for exceptions to this rule, they do not
+appear to be necessary based on the designs. Implementing our dark mode, then,
+should allow component authors to write components _once_, using semantic color
+names, that will automatically switch between their light and dark mode
+counterparts.
+
+### Implementation
+
+We will switch our color names defined in the tailwind configuration to use
+semantic names, for example replacing "pink" with "primary" and "yellow" with
+"complementary". Instead of hardcoding these colors in the Tailwind
+configuration, the tailwind configuration will reference CSS variables defined
+in our root css file. The value of the CSS variables will be switched based on a
+dark mode CSS class added to the HTML root when dark mode is enabled.
+
+Tailwind's built in `dark:` modifier can be used for any styles which need to
+override the default behavior or add dark-mode specific styles beyond the core
+palette swap.
+
+Here are pretend, simplified examples of this setup.
+
+In our primary CSS file, we setup two lists of CSS variables with different
+values when the `.dark-mode` class is present:
+
+```css
+:root {
+  --color-primary: black;
+  --color-secondary: white;
+}
+
+.dark-mode {
+  --color-primary: white;
+  --color-secondary: black;
+}
```

Comment:
Could potentially need to be more complex here, if we want to avoid a light-mode flash for users manually selecting dark mode.

The selector would need to be something like... `.dark-mode OR (prefers-color-scheme: dark AND NOT .light-mode)`, where `.light-mode` is only ever set for users who have manually toggled to light mode from `prefers-color-scheme: dark`.

An inverted selector needed for the default-light mode state as well... `.light-mode OR (prefers-color-scheme: light AND NOT .dark-mode)`.

I think this makes the tailwind darkmode prefix logic a bit complicated too :thinking:

Reply:
My original thought was that the `prefers-color-scheme` media query would be read in JS and used to set this css class. 

For example, a first-time user visits the site, chooses "dark mode" from the UI. On subsequent SSR visits their "dark mode" selection is read from the cookie and the class is added on the server. 

If we do need more advanced tailwind prefix logic, though, they have decent functionality to make it less clunky:

https://tailwindcss.com/docs/dark-mode#using-multiple-selectors

Reply:
Reading it in JS would cause a flash on SSR though for users who's system-level preference is not the default, right? Or is the server able to read that media query for the client? The benefit of using the media query is that it entirely divorces the implementation from JS, meaning the styles the user sees are immediately reflective of their environment and choice (both), without any delay. SSR only comes into play to set a choice that overrides the media query.

We have this issue all over the place with JS-dependent media query options (where we use `v-if="breakpoint == 'md'"` type logic rather than CSS), and it creates rather messy JavaScript and our unfortunate reliance on cookies to SSR the correct breakpoint _on subsequent page loads_ but never the first (except for when we get lucky and the UA indicates a relevant platform, which might not continue to be reliable into the future with UA hints).

Reply:
Updated, and included a strategy for bypassing the cache when this new cookie is present.